python-package/xgboost/spark/core.py

```diff
@@ -0,0 +1,834 @@
+    HasBaseMarginCol,
+)
+
+from pyspark.ml.functions import array_to_vector, vector_to_array
+from pyspark.sql.types import \
+    ArrayType, DoubleType, FloatType, IntegerType, LongType, ShortType
+from pyspark.ml.linalg import VectorUDT
+
+# Put pyspark specific params here, they won't be passed to XGBoost.
+# like `validationIndicatorCol`, `baseMarginCol`
+_pyspark_specific_params = [
+    "featuresCol",
+    "labelCol",
+    "weightCol",
+    "rawPredictionCol",
+    "predictionCol",
+    "probabilityCol",
+    "validationIndicatorCol",
+    "baseMarginCol",
+    "arbitraryParamsDict",
+    "force_repartition",
+    "num_workers",
+    "use_gpu",
+    "feature_types",
+    "feature_names",
+]
+
+_non_booster_params = [
+    "missing",
+    "n_estimators",
+]
+
+_pyspark_param_alias_map = {
+    "features_col": "featuresCol",
+    "label_col": "labelCol",
+    "weight_col": "weightCol",
+    "raw_prediction_ol": "rawPredictionCol",
+    "prediction_col": "predictionCol",
+    "probability_col": "probabilityCol",
+    "validation_indicator_col": "validationIndicatorCol",
+    "base_margin_col": "baseMarginCol",
+}
+
+_unsupported_xgb_params = [
+    "gpu_id",  # we have "use_gpu" pyspark param instead.
+    "enable_categorical",  # TODO: support this
+    "use_label_encoder",
+    "n_jobs",  # Do not allow user to set it, will use `spark.task.cpus` value instead.
+    "nthread",  # Ditto
+]
+
+_unsupported_fit_params = {
+    "sample_weight",  # Supported by spark param weightCol
+    # Supported by spark param weightCol # and validationIndicatorCol
+    "eval_set",
+    "sample_weight_eval_set",
+    "base_margin",  # Supported by spark param baseMarginCol
+}
+
+_unsupported_predict_params = {
+    # for classification, we can use rawPrediction as margin
+    "output_margin",
+    "validate_features",  # TODO
+    "base_margin",  # Use pyspark baseMarginCol param instead.
+}
+
+
+class _XgboostParams(
+    HasFeaturesCol,
+    HasLabelCol,
+    HasWeightCol,
+    HasPredictionCol,
+    HasValidationIndicatorCol,
+    HasArbitraryParamsDict,
+    HasBaseMarginCol,
+):
+    num_workers = Param(
+        Params._dummy(),
+        "num_workers",
+        "The number of XGBoost workers. Each XGBoost worker corresponds to one spark task.",
+        TypeConverters.toInt,
+    )
+    use_gpu = Param(
+        Params._dummy(),
+        "use_gpu",
+        "A boolean variable. Set use_gpu=true if the executors "
+        + "are running on GPU instances. Currently, only one GPU per task is supported.",
+    )
+    force_repartition = Param(
+        Params._dummy(),
+        "force_repartition",
+        "A boolean variable. Set force_repartition=true if you "
+        + "want to force the input dataset to be repartitioned before XGBoost training."
+        + "Note: The auto repartitioning judgement is not fully accurate, so it is recommended"
+        + "to have force_repartition be True.",
+    )
+    feature_names = Param(
```

Comment:
I am not sure if we need to define this parameter on the pyspark side, since both "feature_names" and "feature_types" are already in xgboost.

Reply:
`feature_names` is not in xgboost.sklearn.XGBModel constructor params. So I need define it in pyspark side.
`feature_types` is in xgboost.sklearn.XGBModel constructor params, I remove it from pyspark param list.